spring-kafka/src/main/java/org/springframework/kafka/core/KafkaAdmin.java

```diff
@@ -0,0 +1,205 @@
+ */
+
+package org.springframework.kafka.core;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.CreateTopicsResult;
+import org.apache.kafka.clients.admin.DescribeTopicsResult;
+import org.apache.kafka.clients.admin.NewTopic;
+import org.apache.kafka.clients.admin.TopicDescription;
+import org.apache.kafka.common.acl.AclBinding;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.context.SmartLifecycle;
+
+/**
+ * An admin that delegates to an {@link AdminClient} to create topics defined
+ * in the application context.
+ *
+ * @author Gary Russell
+ * @since 1.3
+ *
+ */
+public class KafkaAdmin implements ApplicationContextAware, SmartLifecycle, DisposableBean {
+
+	private static final int DEFAULT_CLOSE_TIMEOUT = 10;
+
+	private final Log logger = LogFactory.getLog(KafkaAdmin.class);
+
+	private final AdminClient adminClient;
+
+	private boolean running;
+
+	private int phase;
+
+	private boolean autoStartup = true;
+
+	private ApplicationContext applicationContext;
+
+	private int closeTimeout = DEFAULT_CLOSE_TIMEOUT;
+
+	/**
+	 * Create an instance with the provided {@link AdminClient}.
+	 * @param adminClient the client.
+	 */
+	public KafkaAdmin(AdminClient adminClient) {
+		this.adminClient = adminClient;
+	}
+
+	/**
+	 * Create an instance with an {@link AdminClient} created from the supplied
+	 * configuration.
+	 * @param config the configuration.
+	 */
+	public KafkaAdmin(Map<String, Object> config) {
+		this.adminClient = AdminClient.create(config);
+	}
+
+	@Override
+	public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+		this.applicationContext = applicationContext;
+	}
+
+	public AdminClient getAdminClient() {
+		return this.adminClient;
+	}
+
+	public void setRunning(boolean running) {
+		this.running = running;
+	}
+
+	public void setPhase(int phase) {
+		this.phase = phase;
+	}
+
+	public void setAutoStartup(boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	/**
+	 * Set the close timeout in seconds. Defaults to 10 seconds.
+	 * @param closeTimeout the timeout.
+	 */
+	public void setCloseTimeout(int closeTimeout) {
+		this.closeTimeout = closeTimeout;
+	}
+
+	@Override
+	public synchronized void start() {
+		if (!this.running) {
+			if (this.applicationContext != null) {
+				addTopicsIfNeeded(this.applicationContext.getBeansOfType(NewTopic.class, false, false).values());
```

Comment:
Well; I thought `SmartLifecycle` is better because it will allow users to control it themselves. For example; if they want the context to start if the broker is not available, they can set `autoStartup` to `false` and `start()` themselves. With `SmartInitializingSingleton` the context will fail to start.